### PR TITLE
cowardly refuse to delete a directory if it isn't a venv

### PR DESCRIPTION
**Pull request by asottile**
## Contribution checklist:

(also see [CONTRIBUTING.rst](/CONTRIBUTING.rst) for details)

- [x] wrote descriptive pull request text
- [x] added/updated test(s)
- [x] updated/extended the documentation (N/A)
- [x] added relevant [issue keyword](https://help.github.com/articles/closing-issues-using-keywords/)
      in message body (N/A)
- [x] added news fragment in [changelog folder](/docs/changelog)
  * fragment name: `<issue number>.<type>.rst` for example (588.bugfix.rst)
  * `<type>` is must be one of `bugfix`, `feature`, `deprecation`,`breaking`, `doc`, `misc`
  * if PR has no issue: consider creating one first or change it to the PR number after creating the PR
  * "sign" fragment with "by :user:`<your username>`"
  * please use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files - by :user:`superuser`."
  * also see [examples](/docs/changelog/examples.rst)
- [x] added yourself to `CONTRIBUTORS` (preserving alphabetical order)

### CODE REVIEW

File: src/tox/venv.py [hunk 1]
```diff
@@ -694,6 +694,28 @@ def tox_testenv_create(venv, action):
...
+        not venv.path.exists()
+        # does exist, but it's empty => OK
+        or not dir_items
+        # it exists and we're on windows with Lib and Scripts => OK
+        or (INFO.IS_WIN and dir_items > {"Scripts", "Lib"})
+        # non-windows, with lib and bin => OK
+        or dir_items > {"bin", "lib"}
+    ):
+        venv.status = "error"
+        reporter.error("cowardly refusing to delete `envdir`: {}".format(venv.path))
```
Comment (obestwalter) [inline src/tox/venv.py]: How about explaining why we are cowardly refusing? E.g. add 

"- directory contents do not look like a virtual environment."

Comment (asottile) [reply src/tox/venv.py]: updated!

```console
$ tox --devenv docs
py create: /tmp/tox/docs
ERROR: cowardly refusing to delete `envdir` (it does not look like a virtualenv): /tmp/tox/docs
___________________________________ summary ____________________________________
ERROR:   py: error
```

**Comment by asottile** [comment]
since AP's python is still broken:

```console
$ git rev-parse HEAD
3eb1896a98f28fbc5b36837da54faa20f46b850d
$ tox -e fix_lint
fix_lint create: /tmp/tox/.tox/fix_lint
fix_lint installdeps: pre-commit >= 1.14.4, < 2
fix_lint installed: aspy.yaml==1.3.0,cfgv==2.0.0,identify==1.4.3,importlib-metadata==0.17,nodeenv==1.3.3,pre-commit==1.17.0,PyYAML==5.1.1,six==1.12.0,toml==0.10.0,virtualenv==16.6.0,zipp==0.5.1
fix_lint run-test-pre: PYTHONHASHSEED='3597702364'
fix_lint run-test: commands[0] | pre-commit run --all-files --show-diff-on-failure
[INFO] Installing environment for https://github.com/asottile/seed-isort-config.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/mirrors-isort.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/asottile/pyupgrade.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
black....................................................................Passed
blacken-docs.............................................................Passed
seed isort known_third_party.............................................Passed
isort....................................................................Passed
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check Yaml...............................................................Passed
Debug Statements (Python)................................................Passed
Flake8...................................................................Passed
pyupgrade................................................................Passed
rst ``code`` is two backticks............................................Passed
fix_lint run-test: commands[1] | python -c 'import pathlib; print("hint: run {} install to add checks as pre-commit hook".format(pathlib.Path(r"/tmp/tox/.tox/fix_lint") / "bin" / "pre-commit"))'
hint: run /tmp/tox/.tox/fix_lint/bin/pre-commit install to add checks as pre-commit hook
___________________________________ summary ____________________________________
  fix_lint: commands succeeded
  congratulations :)
```

**Comment by westover** [comment]
How do I override this feature?

**Comment by asottile** [comment]
beyond `rm -rf path/to/thing`?  is there a virtualenv type that's not covered here?

**Comment by westover** [comment]
Ended up doing `rm -rf .tox` as part of my CI which seems to alleviate the issue.

**Comment by asottile** [comment]
Do you have a failing output? I'd be interested in what case you're hitting?

**Comment by westover** [comment]
The output isn't very illuminating but here it is:
`/mnt/jenkins/workspace/test-ubermon/env/bin/tox
GLOB sdist-make: /mnt/jenkins/workspace/test-ubermon/setup.py
py27 create: /mnt/jenkins/workspace/test-ubermon/.tox/py27
ERROR: cowardly refusing to delete `envdir` (it does not look like a virtualenv): /mnt/jenkins/workspace/test-ubermon/.tox/py27
___________________________________ summary ____________________________________
ERROR:   py27: error
ERROR:   py34: undefined`

**Comment by asottile** [comment]
interesting!  what version of `tox` are you using and are there any plugins at play? -- if you `ls -al .tox/py27` right before the error occurs can you show me the contents?

**Comment by westover** [comment]
This .tox directory was created by version 2.5 I believe. And then I ran the latest version in the same directory. Unfortunately I could never reproduce the error locally and that directory is long gone at this point

**Comment by asottile** [comment]
wompwomp oh well :( -- if it crops up again let us know! thanks

**Comment by ssbarnea** [comment]
Something is fishy I recently encountered this with env name `dist`, was me unlucky to pick that env name? https://travis-ci.com/pycontribs/jira/jobs/213724388
```
dist create: /home/travis/build/pycontribs/jira/.tox/dist
ERROR: cowardly refusing to delete `envdir` (it does not look like a virtualenv): /home/travis/build
```

**Comment by asottile** [comment]
`./dist` is commonly used by the outputs of `python setup.py $cmd` (sdist, bdist_wheel, etc.) -- the error in that case is correct as tox would have blown away those artifacts

**Comment by asottile** [comment]
Oh hmm .tox/dist I don't know about though :thinking:

**Comment by asottile** [comment]
Ah yes, that's [`distdir`](https://tox.readthedocs.io/en/latest/config.html#conf-distdir)

**Pull request merged**